homeassistant/components/motioneye/__init__.py

```diff
@@ -109,13 +144,85 @@ def _add_camera(
 ) -> None:
     """Add a motionEye camera to hass."""
 
-    device_registry.async_get_or_create(
+    def _is_recognized_web_hook(url: str) -> bool:
+        """Determine whether this integration set a web hook."""
+        return f"{WEB_HOOK_SENTINEL_KEY}={WEB_HOOK_SENTINEL_VALUE}" in url
+
+    def _set_webhook(
+        url: str,
+        key_url: str,
+        key_method: str,
+        key_enabled: str,
+        camera: dict[str, Any],
+    ) -> bool:
+        """Set a web hook."""
+        if (
+            entry.options.get(
+                CONF_WEBHOOK_SET_OVERWRITE,
+                DEFAULT_WEBHOOK_SET_OVERWRITE,
+            )
+            or not camera.get(key_url)
+            or _is_recognized_web_hook(camera[key_url])
+        ) and (
+            not camera.get(key_enabled, False)
+            or camera.get(key_method) != KEY_HTTP_METHOD_GET
+            or camera.get(key_url) != url
+        ):
+            camera[key_enabled] = True
+            camera[key_method] = KEY_HTTP_METHOD_GET
+            camera[key_url] = url
+            return True
+        return False
+
+    def _build_url(base: str, keys: list[str]) -> str:
+        """Build a motionEye webhook URL."""
+
+        return (
+            base
+            + "?"
+            + urlencode(
+                {
+                    **{k: KEY_WEB_HOOK_CONVERSION_SPECIFIERS[k] for k in sorted(keys)},
+                    WEB_HOOK_SENTINEL_KEY: WEB_HOOK_SENTINEL_VALUE,
+                },
+                safe="%{}",
+            )
```

Comment:
We should not construct URLs like this. Instead, use a library to handle that. For example, we have YARL available by default.

Reply:
With the context that this URL is not being fetched here, but rather is being passed to motionEye in JSON -- for motionEye to (significantly) string manipulate and then fetch:

I need the query string to say URL unencoded, i.e. I need to permit raw `[%{}]` characters in the result. YARL automatically encodes all query string parameters and I cannot find any way to work around that.

I've very slightly modified this so I'm using `urljoin` rather than a naked string join, but you may still be unhappy with it. If you have an alternative library recommendation (or any other ideas), I am very happy to dig into this further.

Reply:
Sounds weird and all, as if this is an URL, this should be URL encoded. If motionEye can't handle that correctly; it should be reported and fixed.

Constructing URLs manually like this, is IMHO something that should be avoided.

That said, if there is no other way at this point 🤷

Reply:
Yeah. Personally, I think it's mostly okay, if we think about this not as a URL but a URL _template_. It's not being fetched here, and motionEye does string manipulation to this value _anyway_ prior to fetching. From a motionEye server perspective I wouldn't expect a user to provide a URL encoded value in the UI for the webhook (esp when it will be doing string replacements on the raw value), so I think the same is reasonable for the API.

Furthermore, depending on a change getting committed to motionEye itself is not practical IMO, as experience suggests that would take a very (very) long time :-(